### PR TITLE
Add lcos::local::one_element_channel

### PR DESCRIPTION
**Pull request by hkaiser**
- refactored local channel implementation
- channel::set is now optionally asynchronous
- adding asynchronous iterator support to channel

### CODE REVIEW

File: hpx/lcos/local/channel.hpp [hunk 1]
```diff
@@ -209,10 +222,264 @@ namespace hpx { namespace lcos { namespace local
...
+            {}
+
+            template <typename T1, typename Lock>
+            hpx::future<void> push(T1 && val, Lock& l)
+            {
+                if (!empty_)
+                {
+                    if (push_active_)
+                    {
+                        // error!
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: Can we return an exceptional future here?

File: hpx/lcos/local/channel.hpp [hunk 2]
```diff
@@ -209,10 +222,264 @@ namespace hpx { namespace lcos { namespace local
...
+            }
+
+            template <typename Lock>
+            hpx::future<T> pop(Lock& l)
+            {
+                if (empty_)
+                {
+                    if (pop_active_)
+                    {
+                        // error!
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: Can we return an exceptional future here?

File: hpx/lcos/local/channel.hpp [hunk 3]
```diff
@@ -209,10 +222,264 @@ namespace hpx { namespace lcos { namespace local
...
+                {
+                    util::scoped_unlock<Lock> ul(l);
+                    pop_();                          // trigger waiting pop
+                }
+                return hpx::make_ready_future();
+            }
+
+            void cancel(boost::exception_ptr const& e)
+            {
+                if (pop_active_)
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: this should be done inside the lock as well to avoid races.

File: hpx/lcos/local/channel.hpp [hunk 4]
```diff
@@ -267,174 +534,293 @@ namespace hpx { namespace lcos { namespace local
...
-        }
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename T>
+    class receive_channel : protected detail::channel_base<T>
+    {
+        typedef detail::channel_base<T> base_type;
 
     private:
         friend class channel_iterator<T>;
+        friend class receive_channel<T>;
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: Why does it need to be friends with itself?

File: hpx/lcos/local/channel.hpp [hunk 5]
```diff
@@ -494,169 +880,199 @@ namespace hpx { namespace lcos { namespace local
...
 
-    private:
-        boost::intrusive_ptr<detail::channel_base<util::unused_type> > channel_;
+    public:
+        receive_channel(channel<void> const& c)
+          : base_type(c.get_channel_impl())
+        {}
+        receive_channel(one_element_channel<void> const& c)
+          : base_type(c.get_channel_impl())
+        {}
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: wouldn't a constructor taking a `channel_base const&` be enough?

File: hpx/lcos/local/channel.hpp [hunk 6]
```diff
@@ -267,174 +534,293 @@ namespace hpx { namespace lcos { namespace local
...
     {
+        typedef detail::channel_base<T> base_type;
+
     public:
         send_channel(channel<T> const& c)
-          : channel_(c.channel_)
+          : base_type(c.get_channel_impl())
+        {}
+        send_channel(one_element_channel<T> const& c)
+          : base_type(c.get_channel_impl())
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: wouldn't a constructor taking a `channel_base const&` be enough?

File: hpx/lcos/local/channel.hpp [hunk 7]
```diff
@@ -267,174 +534,293 @@ namespace hpx { namespace lcos { namespace local
...
 
-    private:
-        boost::intrusive_ptr<detail::channel_base<T> > channel_;
+    public:
+        receive_channel(channel<T> const& c)
+          : base_type(c.get_channel_impl())
+        {}
+        receive_channel(one_element_channel<T> const& c)
+          : base_type(c.get_channel_impl())
+        {}
```
Comment (sithhell) [inline hpx/lcos/local/channel.hpp]: wouldn't a constructor taking a `channel_base const&` be enough?

File: hpx/lcos/local/channel.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/lcos/local/channel.hpp]: Right, that was the plan ;)

File: hpx/lcos/local/channel.hpp (hunk 3 above)
Comment (hkaiser) [inline hpx/lcos/local/channel.hpp]: This is called with while the lock is being held. I'll make it explicit.

File: hpx/lcos/local/channel.hpp (hunk 7 above)
Comment (hkaiser) [inline hpx/lcos/local/channel.hpp]: No for two reasons:

- in that case a `receive_channel` would become constructible from a `send_channel`, for instance
- all channel classes are non-publicly derived  (using `protected` derivation) from `channel_base`

File: hpx/lcos/local/channel.hpp (hunk 4 above)
Comment (hkaiser) [inline hpx/lcos/local/channel.hpp]: Right, copy&paste error.

**Comment by sithhell** [comment]
Hmm, I'm not a big fan of that, that makes a little awkward to extend.
Could we do it the other way around, that is have an explicit operator in
the various channels?

Am 18.05.2017 3:55 nachm. schrieb "Hartmut Kaiser" <notifications@github.com

**Comment by hkaiser** [comment]
> Hmm, I'm not a big fan of that

Of what? Could you elaborate, please?

**Comment by sithhell** [comment]
>> Hmm, I'm not a big fan of that
>
> Of what? Could you elaborate, please?

That we essentially have to list all channel implementations in the ctor of `receive_channel` and `send_channel`. Guess it is not something blocking right now, still a little nuisance,

**Comment by hkaiser** [comment]
> That we essentially have to list all channel implementations in the ctor of  receive_channel  and  send_channel . Guess it is not something blocking right now, still a little nuisance,

I don't have a better solution. Besides, being explicit is always better than relying on implicit conversions.

**Pull request merged**